components/org.wso2.carbon.identity.query.saml/src/main/java/org/wso2/carbon/identity/query/saml/util/SAMLQueryRequestUtil.java

```diff
@@ -0,0 +1,509 @@
+
+        AuthnStatement authStmt = new AuthnStatementBuilder().buildObject();
+        authStmt.setAuthnInstant(new DateTime());
+
+        AuthnContext authContext = new AuthnContextBuilder().buildObject();
+        AuthnContextClassRef authCtxClassRef = new AuthnContextClassRefBuilder().buildObject();
+        authCtxClassRef.setAuthnContextClassRef(AuthnContext.PASSWORD_AUTHN_CTX);
+        authContext.setAuthnContextClassRef(authCtxClassRef);
+        authStmt.setAuthnContext(authContext);
+        samlAssertion.getAuthnStatements().add(authStmt);
+
+        if (claims != null) {
+            samlAssertion.getAttributeStatements().add(buildAttributeStatement(claims));
+        }
+
+        AudienceRestriction audienceRestriction = new AudienceRestrictionBuilder().buildObject();
+        Audience issuerAudience = new AudienceBuilder().buildObject();
+        issuerAudience.setAudienceURI(ssoIdPConfigs.getIssuer());
+        audienceRestriction.getAudiences().add(issuerAudience);
+        if (ssoIdPConfigs.getRequestedAudiences() != null) {
+            for (String requestedAudience : ssoIdPConfigs.getRequestedAudiences()) {
+                Audience audience = new AudienceBuilder().buildObject();
+                audience.setAudienceURI(requestedAudience);
+                audienceRestriction.getAudiences().add(audience);
+            }
+        }
+
+        Conditions conditions = new ConditionsBuilder().buildObject();
+        conditions.setNotBefore(currentTime);
+        conditions.setNotOnOrAfter(notOnOrAfter);
+        conditions.getAudienceRestrictions().add(audienceRestriction);
+        samlAssertion.setConditions(conditions);
+
+        if (ssoIdPConfigs.isDoSignAssertions()) {
+
+            try {
+                OpenSAML3Util.setSignature(samlAssertion, ssoIdPConfigs.getSigningAlgorithmUri(), ssoIdPConfigs
+                        .getDigestAlgorithmUri(), new SignKeyDataHolder(tenantDomain));
+            } catch (IdentityException e) {
+                log.error("Unable to set signature to the Assertion", e);
+                throw new IdentitySAML2QueryException("Unable to set signature to the Assertion");
+            }
+        }
+
+        return samlAssertion;
+    }
+
+
+    /**
+     * This method is used to build Attribute Statement including user attributes
+     *
+     * @param claims List of requested claims
+     * @return AttributeStatement set of attributes contain inside attribute statement
+     * @throws  IdentitySAML2QueryException If unable to filter attributes from Map
+     */
+
+    public static AttributeStatement buildAttributeStatement(Map<String, String> claims) throws IdentitySAML2QueryException {
+        AttributeStatement attStmt = null;
+        if (claims != null) {
+            attStmt = new AttributeStatementBuilder().buildObject();
+            Iterator<String> iterator = claims.keySet().iterator();
+
+            for (int i = 0; i < claims.size(); i++) {
```

Comment:
Wouldn't it be better if we use if we get the entrySet of the map and iterate over it? This will prevent us from doing a map lookup in each iteration.

Ref: http://stackoverflow.com/questions/46898/how-to-efficiently-iterate-over-each-entry-in-a-map